### PR TITLE
Cargo: Add support for declared authors

### PR DESCRIPTION
**Pull request by oheger-bosch**
Extract information about the authors of a package from the Cargo.toml
definition file or from the metadata returned by cargo.

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/synthetic/cargo-subcrate/Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 [package]
 name = "lib"
 version = "0.1.0"
-authors = ["anon <anon@example.org>"]
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/cargo-subcrate/Cargo.toml:4]: Why was this removed? To test the implementation if *no* author is given?

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 2]
```diff
@@ -231,6 +231,19 @@ class Cargo(
         return null
     }
 
+    /**
+     * Extract information about authors from the given [node] with package metadata.
+     */
+    private fun parseAuthors(node: JsonNode?): SortedSet<String> =
+        node?.mapNotNullTo(sortedSetOf()) { authorNode -> parseAuthorString(authorNode.textValue()) } ?: sortedSetOf()
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:238]: Nit: For such one-liners we usually stick to using the default `it` in the lambda.

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 3]
```diff
@@ -283,11 +296,13 @@ class Cargo(
         }.let { it.copy(id = it.id.copy(type = managerName)) }
 
         val homepageUrl = pkgDefinition.getString("package.homepage").orEmpty()
+        val authors = pkgDefinition.getList<String>("package.authors", emptyList())
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:299]: Style nit: Maybe move `<String>` to `emptyList()`?

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 4]
```diff
@@ -231,6 +231,19 @@ class Cargo(
...
+     */
+    private fun parseAuthors(node: JsonNode?): SortedSet<String> =
+        node?.mapNotNullTo(sortedSetOf()) { authorNode -> parseAuthorString(authorNode.textValue()) } ?: sortedSetOf()
+
+    /**
+     * Parse an author string of the form "author <email>". Extract the author name or return *null* if this is
+     * not possible.
+     */
+    private fun parseAuthorString(author: String?): String? =
+        author?.substringBefore('<')?.trim()?.ifEmpty { null }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:245]: Given how often we have this function now, how about adding it as a top-level function to `PackageManager.kt`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/funTest/assets/projects/synthetic/cargo-subcrate/Cargo.toml (hunk 1 above)
Comment (oheger-bosch) [reply analyzer/src/funTest/assets/projects/synthetic/cargo-subcrate/Cargo.toml:4]: Yes, exactly for this reason.

File: analyzer/src/main/kotlin/managers/Cargo.kt (hunk 4 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/Cargo.kt:245]: Yes, this would probably make sense. Maybe in a separate PR?

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Cargo.kt:245]: Or in a new follow-up commit in this PR. Your choice.

Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/Cargo.kt]: `PackageManager.kt` now offers a central function for parsing author strings. The concrete package managers have been adjusted to use the new function if applicable in separate commits.

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 5]
```diff
@@ -28,6 +28,7 @@ import java.util.SortedSet
 
 import org.ossreviewtoolkit.analyzer.AbstractPackageManagerFactory
 import org.ossreviewtoolkit.analyzer.PackageManager
+import org.ossreviewtoolkit.analyzer.parseAuthorString
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:31]: Looks like now `@Suppress("TooManyFunctions")` is not needed anymore on this class. So maybe first introduce the global function and take it right away into use for Cargo instead of adding and removing the annotation back and forth?

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 6]
```diff
@@ -204,21 +205,12 @@ open class Npm(
             json["author"]?.let { authorNode ->
                 when {
                     authorNode.isObject -> authorNode["name"]?.textValue()
-                    authorNode.isTextual -> parseAuthorString(authorNode.textValue())
+                    authorNode.isTextual -> parseAuthorString(authorNode.textValue(), '<', '(')
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Npm.kt:208]: Nit: I think it'd be fine to squash all commits that start to make use of the global function.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]

**Review by sschuberth — APPROVED** [review]
